### PR TITLE
Add file system storage and push command

### PR DESCRIPTION
**Pull request by Sifflex**
## Purpose

The purpose of ralph is to be able to store archives in many formats, and the filesystem one was lacking.

## Proposal
- [x] add a FileSystem storage
- [x] add a `push` command to interact implemented backends

Related to #16

### CODE REVIEW

File: src/ralph/backends/storage/fs.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+class FSStorage(HistoryMixin, BaseStorage):
+    """FileSystem storage backend"""
+
+    name = "fs"
+
+    _fspath = "fs_archives/"
+
+    def __init__(self):
+        if not os.path.isdir(self._fspath) and not os.path.exists(self._fspath):
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:22]: Use the `pathlib` module instead of `os.path`.

File: src/ralph/backends/storage/fs.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            logger.debug(
+                "No existing FileSystemStorage, creating a new one at %s", self._fspath
+            )
+            os.mkdir(self._fspath)
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:26]: You should use a recursive path creation instead in case intermediary folders do not exist.

File: src/ralph/backends/storage/fs.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        logger.debug("FileSystemStorage's path: %s", self._fspath)
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:27]: don't forget to set the `self._path` attribute.

File: src/ralph/backends/storage/fs.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            logger.error("%s does not exist", path)
+            raise FileNotFoundError()
+        return path
+
+    def _details(self, name):
+        """Get name archive details"""
+        path = self._get_filepath(name)
+        details = {}
+        details["filename"] = name
+        details["size"] = os.path.getsize(path)
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:42]: Maybe add the creation date?

File: src/ralph/backends/storage/fs.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+        logger.debug("Getting archive: %s", name)
+
+        path = self._get_filepath(name)
+        file = open(path, "rb")
+        chunk = file.read(chunk_size)
+        while chunk:
+            sys.stdout.buffer.write(chunk)
+            chunk = file.read(chunk_size)
+        file.close()
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:81]: Use the `with` statement while opening files and also use the new walrus operator when you can!

```suggestion
        with open(path, "rb") as file:
            while chunk := file.read(chunk_size):
                sys.stdout.buffer.write(chunk)
```

File: src/ralph/backends/storage/fs.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        if os.path.isfile(path) and not overwrite:
+            logger.error("%s already exists and overwrite not specified", name)
+            raise FileExistsError()
+
+        file = open(path, "w")
+        chunk = sys.stdin.read(chunk_size)
+        while chunk:
+            file.write(chunk)
+            chunk = sys.stdin.read(chunk_size)
+        file.close()
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:114]: Apply the same file opening pattern here (`with` statement and walrus operator).

File: src/ralph/backends/storage/fs.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        # Archive is supposed to have been fully created, add a new entry to
+        # the history.
+        self.append_to_history(
+            {
+                "backend": self.name,
+                "command": "push",
+                "id": name,
+                "filename": details.get("filename"),
+                "size": details.get("size"),
+                "fetched_at": datetime.datetime.now(
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:126]:
```suggestion
                "pushed_at": datetime.datetime.now(
```

**Review by jmaupetit — COMMENTED** [review]
That's a pretty good one! :muscle: 



Three remarks:

* Use the `pathlib` module everywhere you need to manipulate a path, that the way we must do it now,
* Add tests for the `FSStorage` backend :pray: 
* Fix your commit messages to comply with our standards

Great work!

File: tests/backends/storage/test_fs.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    posix_path = Path(path)
+    for file in posix_path.iterdir():
+        file.unlink()
+    if path.endswith("/"):
+        path = path[:-1]
+    path = path.split("/")
+    while len(path) > 0:
+        new_path = "/".join(path)
+        Path(new_path).rmdir()
+        path.pop()
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:23]: Use pyfakefs instead (see examples using the `fs` fixture in other tests).

File: tests/test_cli.py [hunk 9]
```diff
@@ -208,3 +241,75 @@ def mock_list(this, details=False, new=False):
...
+    result = runner.invoke(cli, ["list", "-b", "fs"])
+    assert result.exit_code == 0
+    assert "Configured fs backend contains no archive" in result.output
+
+
+def test_push_command_with_fs_backend():
+    """Test the push command using the FS backend"""
+
+    runner = CliRunner()
+    result = runner.invoke(cli, input="echo test | bin/ralph push -b fs file1")
```
Comment (jmaupetit) [inline tests/test_cli.py:311]:
```suggestion
    result = runner.invoke(
        cli,
        [
            "push",
            "-b",
            "fs",
            "file1",
        ],
        input="test"
    )
```

Comment (Sifflex) [reply tests/test_cli.py]: Thank you for the help, it works

File: src/ralph/backends/storage/fs.py (hunk 7 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py]: Thanks for noticing it

File: src/ralph/backends/storage/fs.py (hunk 6 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py]: I didn't know about that walrus operator, it's fixed now

File: src/ralph/backends/storage/fs.py (hunk 4 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py]: Fixed

File: src/ralph/backends/storage/fs.py (hunk 3 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py]: I transformed self._fspath into self._path, is that what you meant ?

File: src/ralph/backends/storage/fs.py (hunk 2 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py]: Done!

File: src/ralph/backends/storage/fs.py (hunk 1 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py]: Done!

File: src/ralph/backends/storage/fs.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class FSStorage(HistoryMixin, BaseStorage):
+    """FileSystem storage backend"""
+
+    name = "fs"
+
+    _fspath = "fs_archives/"
+
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py]: Now it is, thanks

File: tests/backends/storage/test_fs.py (hunk 8 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py]: It is now used in all tests that need it but one, thank you

File: src/ralph/backends/storage/fs.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        )
+
+    def write(self, name, chunk_size=4096, overwrite=False):
+        """Write content to the `name` target"""
+
+        logger.debug("Creating archive: %s", name)
+
+        file_path = self._path.joinpath(name)
+        if file_path.is_file() and not overwrite:
+            logger.error("%s already exists and overwrite not specified", name)
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:115]:
```suggestion
            logger.error("%s already exists and overwrite is not allowed", name)
```

File: src/ralph/backends/storage/fs.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+            raise FileExistsError()
+        file_path = str(file_path)
+
+        file = open(file_path, "w")
+        with open(file_path, "w") as file:
+            while chunk := sys.stdin.read(chunk_size):
+                file.write(chunk)
+            file.close()
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:123]:
```suggestion
        with file_path.open("wb") as file:
            while chunk := sys.stdin.read(chunk_size):
                file.write(chunk)
```

File: src/ralph/defaults.py [hunk 13]
```diff
@@ -35,3 +36,4 @@ class StorageBackends(Enum):
 DEFAULT_GELF_PARSER_CHUNCK_SIZE = 5000
 ENVVAR_PREFIX = "RALPH"
 HISTORY_FILE = Path(environ.get("RALPH_HISTORY_FILE", APP_DIR / "history.json"))
+FS_STORAGE_DEFAULT_PATH = "fs_archives/"
```
Comment (jmaupetit) [inline src/ralph/defaults.py:39]:
```suggestion
FS_STORAGE_DEFAULT_PATH = "./archives"
```

File: src/ralph/backends/storage/fs.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+            logger.debug(
+                "No existing FileSystemStorage, creating a new one at %s", self._path
+            )
+            self._path.mkdir(parents=True)
+
+        logger.debug("FileSystemStorage's path: %s", self._path)
+
+    def _get_filepath(self, name):
+        """Return path of the archive in the fsstorage, or throws an exception if not found"""
+        file_path = self._path.joinpath(name)
```
Comment (jmaupetit) [inline src/ralph/backends/storage/fs.py:36]: Do you now the paths join operator? I love this kind of syntactic sugar:

```suggestion
        file_path = self._path / Path(name)
```

File: tests/backends/storage/test_fs.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    # pylint: disable=protected-access
+
+    assert FSStorage.name == "fs"
+
+    storage = FSStorage()
+
+    assert storage._path == Path(FS_STORAGE_DEFAULT_PATH)
+
+    if Path("multi").exists():
+        assert False
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:25]: I don't get this.

File: tests/backends/storage/test_fs.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+    multi_path = "multi/directories/path"
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:27]:
```suggestion
    deep_path = "deep/directories/path"
```

File: tests/backends/storage/test_fs.py [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    # pylint: disable=protected-access
+
+    path = "test_fs/"
+    filename = "some_file"
+    storage = FSStorage(path)
+
+    with pytest.raises(FileNotFoundError):
+        storage._get_filepath(filename)
+
+    open(path + filename, "w").close()
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:47]: Use the `fs` fixture for that:

```suggestion
    fs.create_file(Path(path, filename))
```

File: tests/backends/storage/test_fs.py [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+    getfilepath = storage._get_filepath(filename)
+
+    assert getfilepath == Path(path, filename)
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:51]:
```suggestion
    assert storage._get_filepath(filename) == Path(path, filename)
```

File: tests/backends/storage/test_fs.py [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+# pylint: disable=invalid-name
+# pylint: disable=unused-argument
+def test_fs_url(fs):
+    """Test that the full url of the file can be returned"""
+
+    path = "test_fs/"
+    filename = "some_file"
+    storage = FSStorage(path)
+
+    open(path + filename, "w").close()
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:63]: ditto

File: tests/backends/storage/test_fs.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+    assert storage.url(filename) == "test_fs/some_file"
+
+
+# pylint: disable=invalid-name
+# pylint: disable=unused-argument
+def test_fs_list(fs):
+    """Test archives listing in FSStorage"""
+
+    Path(APP_DIR).mkdir(parents=True)
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:73]:
```suggestion
    fs.create_dir(APP_DIR)
```

File: tests/backends/storage/test_fs.py [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+    path = "test_fs/"
+    filename1 = "file1"
+    filename2 = "file2"
+    storage = FSStorage(path)
+
+    file1 = open(path + filename1, "w")
+    file1.write("content")
+    file1.close()
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:82]:
```suggestion
    fs.create_file(path + filename1, contents="content")
```

File: tests/backends/storage/test_fs.py [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+
+    file2 = open(path + filename2, "w")
+    file2.write("some more content")
+    file2.close()
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:86]:
```suggestion
    fs.create_file(path + filename2, contents="some more content")
```

File: tests/backends/storage/test_fs.py [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+
+    assert isinstance(storage.list(), Iterable)
+    assert isinstance(storage.list(new=True), Iterable)
+    assert isinstance(storage.list(details=True), Iterable)
+
+    simple_list = list(storage.list())
+    assert filename1 in simple_list and filename2 in simple_list
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:93]:
```suggestion
    assert filename1 in simple_list
    assert filename2 in simple_list
```

File: tests/backends/storage/test_fs.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+
+    # Fetch it so it's not new anymore
+    storage.read(filename1)
+
+    new_list = list(storage.list(new=True))
+    assert filename1 not in new_list and filename2 in new_list
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:99]:
```suggestion
    assert filename1 not in new_list
    assert filename2 in new_list
```

File: tests/backends/storage/test_fs.py [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+
+    detail_list = list(storage.list(details=True))
+    assert any(
+        (elt["filename"] == filename1 and elt["size"] == 7) for elt in detail_list
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:103]:
```suggestion
        (archive["filename"] == filename1 and archive["size"] == 7) for archive in detail_list
```

File: tests/backends/storage/test_fs.py [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+    )
+    assert any(
+        (elt["filename"] == filename2 and elt["size"] == 17) for elt in detail_list
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:106]: ditto

File: tests/test_cli.py [hunk 27]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+    """Test the list command using the LDP backend"""
+
+    archive_list = [
+        "file1",
+        "file2",
+    ]
+    archive_list_details = [
+        {
+            "filename": "file1",
+            "modifiedAt": "1604661859",
```
Comment (jmaupetit) [inline tests/test_cli.py:257]: I prefer using snake case (see my previous comment):

```suggestion
            "modified_at": "1604661859",
```

File: tests/test_cli.py [hunk 28]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+            "size": 350,
+        },
+        {
+            "filename": "file2",
+            "createdAt": "1604935848",
```
Comment (jmaupetit) [inline tests/test_cli.py:262]:
```suggestion
            "created_at": "1604935848",
```

File: tests/test_cli.py [hunk 29]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+    result = runner.invoke(cli, ["list", "-b", "fs", "-n"])
+    assert result.exit_code == 0
+    assert "file2" in result.output
+    assert "file1" not in result.output
+
+    # Edge case: stream contains no archive
+    monkeypatch.setattr(FSStorage, "list", lambda this, details, new: ())
+    result = runner.invoke(cli, ["list", "-b", "fs"])
+    assert result.exit_code == 0
+    assert "Configured fs backend contains no archive" in result.output
```
Comment (jmaupetit) [inline tests/test_cli.py:305]: :clap:

File: tests/test_cli.py [hunk 30]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+            "file1",
+        ],
+        input="test content",
+    )
+
+    assert result.exit_code == 0
+
+    test_file = open(file_path)
+    content = test_file.read()
+    test_file.close()
```
Comment (jmaupetit) [inline tests/test_cli.py:333]:
```suggestion
    with file_path.open() as test_file:
        content = test_file.read()
```

File: tests/test_cli.py [hunk 31]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+    assert "Configured fs backend contains no archive" in result.output
+
+
+def test_push_command_with_fs_backend():
+    """Test the push command using the FS backend
+    It is not possible to use pyfakefs here due to a direct use of the cli which cannot be mocked
+    """
+
+    filename = "file1"
+    file_path = FS_STORAGE_DEFAULT_PATH + filename
```
Comment (jmaupetit) [inline tests/test_cli.py:314]: `FS_STORAGE_DEFAULT_PATH` should be a `Path` instance:

```suggestion
    file_path = FS_STORAGE_DEFAULT_PATH / Path(filename)
```

File: tests/test_cli.py [hunk 32]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+
+    assert result.exit_code == 0
+
+    test_file = open(file_path)
+    content = test_file.read()
+    test_file.close()
+
+    assert "test content" in content
+
+    # Try to create the same file without -f, error 1
```
Comment (jmaupetit) [inline tests/test_cli.py:337]:
```suggestion
    # Try to create the same file without -f should raise an error
```

File: tests/test_cli.py [hunk 33]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+        cli,
+        [
+            "push",
+            "-b",
+            "fs",
+            "file1",
+        ],
+        input="different content",
+    )
+    assert result.exit_code == 1
```
Comment (jmaupetit) [inline tests/test_cli.py:349]: Maybe also test the error message?

File: tests/test_cli.py [hunk 34]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+            "file1",
+        ],
+        input="different content",
+    )
+
+    assert result.exit_code == 0
+
+    test_file = open(file_path)
+    content = test_file.read()
+    test_file.close()
```
Comment (jmaupetit) [inline tests/test_cli.py:369]:
```suggestion
    with file_path.open() as test_file:
        content = test_file.read()
```

File: tests/test_cli.py [hunk 35]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+
+    assert "different content" in content
+
+    for archive in Path(FS_STORAGE_DEFAULT_PATH).iterdir():
+        Path(archive).unlink()
+    Path(FS_STORAGE_DEFAULT_PATH).rmdir()
```
Comment (jmaupetit) [inline tests/test_cli.py:375]: You don't need this while using the `fs` fixture.

File: tests/test_cli.py [hunk 36]
```diff
@@ -208,3 +242,134 @@ def mock_list(this, details=False, new=False):
...
+    # Edge case: stream contains no archive
+    monkeypatch.setattr(FSStorage, "list", lambda this, details, new: ())
+    result = runner.invoke(cli, ["list", "-b", "fs"])
+    assert result.exit_code == 0
+    assert "Configured fs backend contains no archive" in result.output
+
+
+def test_push_command_with_fs_backend():
+    """Test the push command using the FS backend
+    It is not possible to use pyfakefs here due to a direct use of the cli which cannot be mocked
```
Comment (jmaupetit) [inline tests/test_cli.py:310]: This is odd.

**Review by jmaupetit — COMMENTED** [review]
This is a really great work! I've taken the time to make a careful review of your work to make your code more pythonic. Keep it up!

File: src/ralph/backends/storage/fs.py [hunk 37]
```diff
@@ -0,0 +1,139 @@
...
+
+
+class FSStorage(HistoryMixin, BaseStorage):
+    """FileSystem storage backend"""
+
+    name = "fs"
+
+    def __init__(self, path=None):
+        if path is None:
+            path = FS_STORAGE_DEFAULT_PATH
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:23]: For some reason this instantiation method does not work when Ralph is called from command line
In tests it works fine though

File: tests/test_cli.py (hunk 36 above)
Comment (Sifflex) [reply tests/test_cli.py:310]: You are right, it was actually possible to use it, it just needed a little workaround that you had used in the test_history_mixin

File: tests/test_cli.py (hunk 35 above)
Comment (Sifflex) [reply tests/test_cli.py:375]: Indeed

File: tests/test_cli.py (hunk 34 above)
Comment (Sifflex) [reply tests/test_cli.py:369]: Thank you, fixed

File: tests/test_cli.py (hunk 33 above)
Comment (Sifflex) [reply tests/test_cli.py:349]: Done

File: tests/test_cli.py (hunk 32 above)
Comment (Sifflex) [reply tests/test_cli.py:337]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 38]
```diff
@@ -0,0 +1,139 @@
...
+            self._path.mkdir(parents=True)
+
+        logger.debug("FileSystemStorage's path: %s", self._path)
+
+    def _get_filepath(self, name):
+        """Return path of the archive in the fsstorage, or throws an exception if not found"""
+        file_path = self._path.joinpath(name)
+        if not file_path.exists():
+            logger.error("%s does not exist", file_path)
+            raise FileNotFoundError()
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:39]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 39]
```diff
@@ -0,0 +1,139 @@
...
+            path = FS_STORAGE_DEFAULT_PATH
+        self._path = Path(path)
+
+        if not self._path.is_dir() and not self._path.exists():
+            logger.debug(
+                "No existing FileSystemStorage, creating a new one at %s", self._path
+            )
+            self._path.mkdir(parents=True)
+
+        logger.debug("FileSystemStorage's path: %s", self._path)
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:32]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 40]
```diff
@@ -0,0 +1,139 @@
...
+
+    def __init__(self, path=None):
+        if path is None:
+            path = FS_STORAGE_DEFAULT_PATH
+        self._path = Path(path)
+
+        if not self._path.is_dir() and not self._path.exists():
+            logger.debug(
+                "No existing FileSystemStorage, creating a new one at %s", self._path
+            )
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:29]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 41]
```diff
@@ -0,0 +1,139 @@
...
+    """FileSystem storage backend"""
+
+    name = "fs"
+
+    def __init__(self, path=None):
+        if path is None:
+            path = FS_STORAGE_DEFAULT_PATH
+        self._path = Path(path)
+
+        if not self._path.is_dir() and not self._path.exists():
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:26]: I forgot to check that, thank you

File: src/ralph/backends/storage/fs.py [hunk 42]
```diff
@@ -0,0 +1,139 @@
...
+        stats = file_path.stat()
+
+        details = {}
+        details["filename"] = name
+        details["size"] = stats.st_size
+        details["modifiedAt"] = datetime.datetime.fromtimestamp(
+            int(stats.st_mtime), tz=datetime.timezone.utc
+        ).isoformat()
+
+        return details
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:50]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 43]
```diff
@@ -0,0 +1,139 @@
...
+
+    def list(self, details=False, new=False):
+        """List files in the storage backend"""
+        archives = [str(archive).rsplit("/")[-1] for archive in self._path.iterdir()]
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:54]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 44]
```diff
@@ -0,0 +1,139 @@
...
+            logger.debug("New archives: %d", len(archives))
+
+        for archive in archives:
+            yield self._details(archive) if details else archive
+
+    def url(self, name):
+        """Get `name` file absolute URL"""
+
+        file_path = self._get_filepath(name)
+        return str(file_path)
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:75]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 45]
```diff
@@ -0,0 +1,139 @@
...
+    def read(self, name, chunk_size=4096):
+        """Read `name` file and stream its content by chunks of a given size"""
+
+        logger.debug("Getting archive: %s", name)
+
+        file_path = str(self._get_filepath(name))
+        with open(file_path, "rb") as file:
+            while chunk := file.read(chunk_size):
+                sys.stdout.buffer.write(chunk)
+            file.close()
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:86]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 46]
```diff
@@ -0,0 +1,139 @@
...
+
+        if not self._path.is_dir() and not self._path.exists():
+            logger.debug(
+                "No existing FileSystemStorage, creating a new one at %s", self._path
+            )
+            self._path.mkdir(parents=True)
+
+        logger.debug("FileSystemStorage's path: %s", self._path)
+
+    def _get_filepath(self, name):
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:29]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 47]
```diff
@@ -0,0 +1,139 @@
...
+        """Return path of the archive in the fsstorage, or throws an exception if not found"""
+        file_path = self._path.joinpath(name)
+        if not file_path.exists():
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:32]: Fixed

File: src/ralph/backends/storage/fs.py [hunk 48]
```diff
@@ -0,0 +1,139 @@
...
+                ).isoformat(),
+            }
+        )
+
+    def write(self, name, chunk_size=4096, overwrite=False):
+        """Write content to the `name` target"""
+
+        logger.debug("Creating archive: %s", name)
+
+        file_path = self._path.joinpath(name)
```
[the remark this answers is not in the document]
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:109]: Fixed

File: src/ralph/backends/storage/fs.py (hunk 11 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:111]: Fixed

File: src/ralph/backends/storage/fs.py (hunk 12 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:119]: Fixed

File: src/ralph/defaults.py (hunk 13 above)
Comment (Sifflex) [reply src/ralph/defaults.py:39]: Fixed

File: src/ralph/backends/storage/fs.py (hunk 14 above)
Comment (Sifflex) [reply src/ralph/backends/storage/fs.py:31]: I didn't know about it, is it specific to pathlib?
Fixed

File: tests/backends/storage/test_fs.py (hunk 15 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py]: Some dead code which had been forgotten

File: tests/backends/storage/test_fs.py (hunk 16 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:24]: Fixed

File: tests/backends/storage/test_fs.py (hunk 17 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:44]: Fixed

File: tests/backends/storage/test_fs.py (hunk 18 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:48]: Fixed

File: tests/backends/storage/test_fs.py (hunk 20 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:70]: Fixed

File: tests/backends/storage/test_fs.py (hunk 21 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:79]: Fixed

File: tests/backends/storage/test_fs.py (hunk 22 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:83]: Fixed

File: tests/backends/storage/test_fs.py (hunk 23 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:90]: Fixed

File: tests/backends/storage/test_fs.py (hunk 24 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:96]: Fixed

File: tests/backends/storage/test_fs.py (hunk 25 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:100]: Fixed

File: tests/backends/storage/test_fs.py (hunk 26 above)
Comment (Sifflex) [reply tests/backends/storage/test_fs.py:103]: Fixed

File: tests/test_cli.py (hunk 27 above)
Comment (Sifflex) [reply tests/test_cli.py:257]: Fixed

File: tests/test_cli.py (hunk 28 above)
Comment (Sifflex) [reply tests/test_cli.py:262]: Fixed

File: tests/test_cli.py (hunk 30 above)
Comment (Sifflex) [reply tests/test_cli.py]: Fixed

File: tests/test_cli.py (hunk 31 above)
Comment (Sifflex) [reply tests/test_cli.py]: Fixed

File: src/ralph/backends/storage/fs.py (hunk 14 above)
Comment (jmaupetit) [reply src/ralph/backends/storage/fs.py]: Yes it is.

File: src/ralph/cli.py [hunk 49]
```diff
@@ -103,12 +104,33 @@ def fetch(backend, archive, **options):
     logger.info("Fetching archive %s from the configured %s backend", archive, backend)
     logger.debug("Backend parameters: %s", options)
 
+    options["fs_path"] = FS_STORAGE_DEFAULT_PATH
```
Comment (jmaupetit) [inline src/ralph/cli.py:107]: You need to fix default value in the CLI argument instead. I've pushed a fix for this.

File: src/ralph/cli.py [hunk 50]
```diff
@@ -103,12 +104,33 @@ def fetch(backend, archive, **options):
...
+@click.option(
+    "-f", "--force", default=False, is_flag=True, help="Overwrite existing file"
+)
+def push(backend, archive, force, **options):
+    """Push an archive to a configured storage backend"""
+
+    logger.info("Pushing archive %s to the configured %s backend", archive, backend)
+    logger.debug("Backend parameters: %s", options)
+
+    options["fs_path"] = FS_STORAGE_DEFAULT_PATH
```
Comment (jmaupetit) [inline src/ralph/cli.py:126]: ditto

File: src/ralph/cli.py [hunk 51]
```diff
@@ -129,6 +151,8 @@ def list_(details, new, backend, **options):
     logger.debug("Fetch details: %s", str(details))
     logger.debug("Backend parameters: %s", options)
 
+    options["fs_path"] = FS_STORAGE_DEFAULT_PATH
```
Comment (jmaupetit) [inline src/ralph/cli.py:154]: ditto

File: tests/backends/storage/test_fs.py [hunk 52]
```diff
@@ -0,0 +1,102 @@
...
+
+    storage = FSStorage()
+
+    assert storage._path == Path(FS_STORAGE_DEFAULT_PATH)
+
+    deep_path = "deep/directories/path"
+
+    storage = FSStorage(deep_path)
+
+    assert storage._path == Path(deep_path)
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:28]: You should also assert that it has been created! And check that it does not raise an error if this path already exists.

File: tests/backends/storage/test_fs.py [hunk 53]
```diff
@@ -0,0 +1,102 @@
...
+    filename = "some_file"
+    storage = FSStorage(path)
+
+    storage._get_filepath(filename)
+    with pytest.raises(FileNotFoundError):
+        storage._get_filepath(filename, strict=True)
+
+    fs.create_file(Path(path, filename))
+
+    assert storage._get_filepath(filename) == Path(path, filename)
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:47]:
```suggestion
    assert storage._get_filepath(filename, strict=True) == Path(path, filename)
```

File: tests/backends/storage/test_fs.py [hunk 54]
```diff
@@ -0,0 +1,102 @@
...
+    """Test that a path that exists can be returned, and throws an exception otherwise"""
+    # pylint: disable=protected-access
+
+    path = "test_fs/"
+    filename = "some_file"
+    storage = FSStorage(path)
+
+    storage._get_filepath(filename)
+    with pytest.raises(FileNotFoundError):
+        storage._get_filepath(filename, strict=True)
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:43]: Also check that it does not raise an error if `strict` is `False`

File: tests/backends/storage/test_fs.py [hunk 55]
```diff
@@ -0,0 +1,102 @@
...
+    fs.create_file(path + filename1, contents="content")
+    fs.create_file(path + filename2, contents="some more content")
+
+    assert isinstance(storage.list(), Iterable)
+    assert isinstance(storage.list(new=True), Iterable)
+    assert isinstance(storage.list(details=True), Iterable)
+
+    simple_list = list(storage.list())
+    assert filename1 in simple_list
+    assert filename2 in simple_list
```
Comment (jmaupetit) [inline tests/backends/storage/test_fs.py:85]:
```suggestion
    assert filename2 in simple_list
    assert len(simple_list) == 2
```

**Review by jmaupetit — COMMENTED** [review]
I only have a few cosmetic comments. I've pushed a fixup [here](https://github.com/openfun/ralph/commit/fa84724c87328dfad203cbabb4666977ba2e48f1), can you:

1. rebase your branch with this fixup,
2. fix your commits ordering (fs backend first and then the push command as tests from the push command depend on the fs backend)

**Review by jmaupetit — APPROVED** [review]
